README.md
# tiny-painter

tiny-painter is a simple, fast and lightweight painting toolkit.

## Usage

```ts
// Create a new painter
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.1-alpha.0 (2022-05-18)


assets/index.50592ad5.js
const p=function(){const r=document.createElement("link").relList;if(r&&r.supports&&r.supports("modulepreload"))return;for(const e of document.querySelectorAll('link[rel="modulepreload"]'))c(e);new MutationObserver(e=>{for(const o of e)if(o.type==="childList")for(const t of o.addedNodes)t.tagName==="LINK"&&t.rel==="modulepreload"&&c(t)}).observe(document,{childList:!0,subtree:!0});function l(e){const o={};return e.integrity&&(o.integrity=e.integrity),e.referrerpolicy&&(o.referrerPolicy=e.referrerpolicy),e.crossorigin==="use-credentials"?o.credentials="include":e.crossorigin==="anonymous"?o.credentials="omit":o.credentials="same-origin",o}function c(e){if(e.ep)return;e.ep=!0;const o=l(e);fetch(e.href,o)}};p();function v(){const n=document.createElement("canvas"),r=document.querySelector("#app");r.style.border="1px solid #000";const{offsetWidth:l,offsetHeight:c,offsetLeft:e,offsetTop:o}=r;n.width=l*window.devicePixelRatio,n.height=c*window.devicePixelRatio,n.style.width=`${l}px`,n.style.height=`${c}px`,r.appendChild(n);const t=n.getContext("2d"),s=[];function a(){t.clearRect(0,0,n.width,n.height)}t.lineWidth=10,t.lineCap="round",t.lineJoin="round";function m(i){if(i.length%2===1){a();for(let d=0;d<i.length-1;d+=2)t.moveTo(...i[d]),t.quadraticCurveTo(...i[d+1],...i[d+2]),t.stroke()}else t.lineTo(...i[s.length-1]),t.stroke()}function u(i){s.push([(i.pageX-e)*window.devicePixelRatio,(i.pageY-o)*window.devicePixelRatio]),s.length===1?t.moveTo(...s[0]):m(s)}function f(){document.removeEventListener("mousemove",u),document.removeEventListener("mouseup",f)}function h(){s.length=0,document.addEventListener("mousemove",u),document.addEventListener("mouseup",f)}n.addEventListener("mousedown",h)}v();


assets/index.7b8328ab.js
const M=function(){const r=document.createElement("link").relList;if(r&&r.supports&&r.supports("modulepreload"))return;for(const e of document.querySelectorAll('link[rel="modulepreload"]'))l(e);new MutationObserver(e=>{for(const t of e)if(t.type==="childList")for(const u of t.addedNodes)u.tagName==="LINK"&&u.rel==="modulepreload"&&l(u)}).observe(document,{childList:!0,subtree:!0});function o(e){const t={};return e.integrity&&(t.integrity=e.integrity),e.referrerpolicy&&(t.referrerPolicy=e.referrerpolicy),e.crossorigin==="use-credentials"?t.credentials="include":e.crossorigin==="anonymous"?t.credentials="omit":t.credentials="same-origin",t}function l(e){if(e.ep)return;e.ep=!0;const t=o(e);fetch(e.href,t)}};M();function w(i,r){Object.keys(r).forEach(o=>{i.style[o]=r[o]})}function g(i=100,r=100){const o=document.createElement("canvas");return o.width=i,o.height=r,w(o,{width:"100%",height:"100%"}),o}function O(){const i=document.querySelector("#app");i.style.border="1px solid #000",i.style.position="relative";const{offsetWidth:r,offsetHeight:o,offsetLeft:l,offsetTop:e}=i,t=r*window.devicePixelRatio,u=o*window.devicePixelRatio,a=g(t,u),f=g(a.width,a.height);w(f,{left:"0px",top:"0px",position:"absolute"}),i.appendChild(a),i.appendChild(f);const d=a.getContext("2d"),c=f.getContext("2d");c.lineWidth=10,c.lineCap="round",c.lineJoin="round",d.lineWidth=10,d.lineCap="round",d.lineJoin="round";const s=[];function y(n){n.clearRect(0,0,n.canvas.width,n.canvas.height)}function x(){y(c),d.moveTo(...s[0]);for(let n=1;n<s.length-1;){const[m,v]=s[n],[C,E]=s[n+1],[b,P]=[(m+C)/2,(v+E)/2];d.quadraticCurveTo(m,v,b,P),n++,d.stroke()}}function p(n){s.push([(n.pageX-l)*window.devicePixelRatio,(n.pageY-e)*window.devicePixelRatio]),c.lineTo(...s[s.length-1]),c.stroke()}function h(){x(),document.removeEventListener("mousemove",p),document.removeEventListener("mouseup",h)}function L(){s.length=0,c.beginPath(),document.addEventListener("mousemove",p),document.addEventListener("mouseup",h)}f.addEventListener("mousedown",L)}O();


assets/index.7c05f569.js
const p=function(){const r=document.createElement("link").relList;if(r&&r.supports&&r.supports("modulepreload"))return;for(const e of document.querySelectorAll('link[rel="modulepreload"]'))c(e);new MutationObserver(e=>{for(const o of e)if(o.type==="childList")for(const t of o.addedNodes)t.tagName==="LINK"&&t.rel==="modulepreload"&&c(t)}).observe(document,{childList:!0,subtree:!0});function l(e){const o={};return e.integrity&&(o.integrity=e.integrity),e.referrerpolicy&&(o.referrerPolicy=e.referrerpolicy),e.crossorigin==="use-credentials"?o.credentials="include":e.crossorigin==="anonymous"?o.credentials="omit":o.credentials="same-origin",o}function c(e){if(e.ep)return;e.ep=!0;const o=l(e);fetch(e.href,o)}};p();function v(){const n=document.createElement("canvas"),r=document.querySelector("#app");r.style.border="1px solid #000";const{offsetWidth:l,offsetHeight:c,offsetLeft:e,offsetTop:o}=r;n.width=l*window.devicePixelRatio,n.height=c*window.devicePixelRatio,n.style.width=`${l}px`,n.style.height=`${c}px`,r.appendChild(n);const t=n.getContext("2d"),s=[];t.lineWidth=10,t.lineCap="round",t.lineJoin="round";function a(){t.clearRect(0,0,n.width,n.height)}function m(i){if(i.length%2===1){a();for(let d=0;d<i.length-1;d+=2)t.moveTo(...i[d]),t.quadraticCurveTo(...i[d+1],...i[d+2]),t.stroke()}else t.lineTo(...i[s.length-1]),t.stroke()}function u(i){s.push([(i.pageX-e)*window.devicePixelRatio,(i.pageY-o)*window.devicePixelRatio]),s.length===1?t.moveTo(...s[0]):m(s)}function f(){document.removeEventListener("mousemove",u),document.removeEventListener("mouseup",f)}function h(){s.length=0,document.addEventListener("mousemove",u),document.addEventListener("mouseup",f)}n.addEventListener("mousedown",h)}v();


assets/index.89fb4632.js
const J=function(){const a=document.createElement("link").relList;if(a&&a.supports&&a.supports("modulepreload"))return;for(const e of document.querySelectorAll('link[rel="modulepreload"]'))x(e);new MutationObserver(e=>{for(const o of e)if(o.type==="childList")for(const v of o.addedNodes)v.tagName==="LINK"&&v.rel==="modulepreload"&&x(v)}).observe(document,{childList:!0,subtree:!0});function s(e){const o={};return e.integrity&&(o.integrity=e.integrity),e.referrerpolicy&&(o.referrerPolicy=e.referrerpolicy),e.crossorigin==="use-credentials"?o.credentials="include":e.crossorigin==="anonymous"?o.credentials="omit":o.credentials="same-origin",o}function x(e){if(e.ep)return;e.ep=!0;const o=s(e);fetch(e.href,o)}};J();function O(c,a){Object.keys(a).forEach(s=>{c.style[s]=a[s]})}function W(c=100,a=100){const s=document.createElement("canvas");return s.width=c,s.height=a,O(s,{width:"100%",height:"100%"}),s}function j(){const c=document.querySelector("#app");c.style.border="1px solid #000",c.style.position="relative";const{offsetWidth:a,offsetHeight:s,offsetLeft:x,offsetTop:e}=c,o=a*window.devicePixelRatio,v=s*window.devicePixelRatio,g=W(o,v),w=W(g.width,g.height);O(w,{left:"0px",top:"0px",position:"absolute"}),c.appendChild(g),c.appendChild(w);const n=g.getContext("2d"),r=w.getContext("2d"),m=30;r.lineWidth=m,r.lineCap="round",r.lineJoin="round",n.lineWidth=m,n.lineCap="round",n.lineJoin="round";let t=[],b=0;function M(i){i.clearRect(0,0,i.canvas.width,i.canvas.height)}function R(){M(r);const[{x:i,y:l}]=t;n.beginPath(),n.moveTo(i,l);for(let d=1;d<t.length-1;){const{x:u,y:p,v:y}=t[d],{x:f,y:h,v:L}=t[d+1],[C,S]=[(u+f)/2,(p+h)/2];n.save();const D=(y+L)/2;n.lineWidth=m/D,n.quadraticCurveTo(u,p,C,S),d++,n.stroke(),n.restore()}}function q(){M(r);const[{x:i,y:l}]=t;n.beginPath(),n.save(),n.lineTo(i,l),n.stroke(),n.restore()}function k(i,l){const{x:d,y:u}=i,{x:p,y}=l,f=Math.abs(d-p)**2,h=Math.abs(u-y)**2;return Math.sqrt(f+h)}function P(){t=[],b=0,r.beginPath()}function E(i){const l={x:(i.pageX-x)*window.devicePixelRatio,y:(i.pageY-e)*window.devicePixelRatio,v:1};t.push(l);const d=Date.now();let u=1;if(b!==0){const f=d-b,h=k(t[t.length-2],t[t.length-1]);u=Math.max(Math.min(h/f,m),1)}l.v=u,b=d,r.save(),r.lineWidth=m/u;const{x:p,y}=t[t.length-1];if(r.lineTo(p,y),r.stroke(),r.restore(),t.length===2){const[{x:f,y:h}]=t,[,{x:L,y:C}]=t;f===L&&h===C&&(q(),P())}}function T(){t.length>0&&R(),document.removeEventListener("mousemove",E),document.removeEventListener("mouseup",T)}function N(){P(),document.addEventListener("mousemove",E),document.addEventListener("mouseup",T)}w.addEventListener("mousedown",N)}j();


src/index.ts
function setStyle(el: HTMLElement, style: Partial<CSSStyleDeclaration>) {
  Object.keys(style).forEach(key => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    el.style[key] = style[key];
  });
}


function createCanvas(width = 100, height = 100) {
  const canvas = document.createElement('canvas');
  canvas.width = width;
  canvas.height = height;
  setStyle(canvas, {
    width: '100%',
    height: '100%',
  });
  return canvas;
}

export interface IPoint {
  x: number;
  y: number;
  v: number;
}

export function main() {
  const container = document.querySelector('#app') as HTMLElement;
  container.style.border = '1px solid #000';
  container.style.position = 'relative';
  const { offsetWidth, offsetHeight, offsetLeft, offsetTop } = container;
  const width = offsetWidth * window.devicePixelRatio;
  const height = offsetHeight * window.devicePixelRatio;
  const canvas = createCanvas(width, height);
  const topCanvas = createCanvas(canvas.width, canvas.height);
  setStyle(topCanvas, {
    left: '0px',
    top: '0px',
    position: 'absolute',
  });

  container.appendChild(canvas);
  container.appendChild(topCanvas);

  const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;
  const topCtx = topCanvas.getContext('2d') as CanvasRenderingContext2D;

  const lineWidth = 30;
  topCtx.lineWidth = lineWidth;
  topCtx.lineCap = 'round';
  topCtx.lineJoin = 'round';
  ctx.lineWidth = lineWidth;
  ctx.lineCap = 'round';
  ctx.lineJoin = 'round';

  let points: IPoint[] = [];
  let startTime = 0;

  function clear(context: CanvasRenderingContext2D) {
    context.clearRect(0, 0, context.canvas.width, context.canvas.height);
  }

  function draw() {
    clear(topCtx);
    const [ { x: sx, y: sy } ] = points;
    ctx.beginPath();
    ctx.moveTo(sx, sy);
    for (let index = 1; index < points.length - 1;) {
      const { x: x1, y: y1, v: v1 } = points[index];
      const { x: x2, y: y2, v: v2 } = points[index + 1];
      const [ x, y ] = [ (x1 + x2) / 2, (y1 + y2) / 2 ];
      ctx.save();
      const v = (v1 + v2) / 2;
      ctx.lineWidth = lineWidth / v;
      ctx.quadraticCurveTo(x1, y1, x, y);
      index++;
      ctx.stroke();
      ctx.restore();
    }
  }

  function drawPoint() {
    clear(topCtx);
    const [ { x: sx, y: sy } ] = points;
    ctx.beginPath();
    ctx.save();
    ctx.lineTo(sx, sy);
    ctx.stroke();
    ctx.restore();
  }

  function distance(p1: IPoint, p2: IPoint) {
    const { x: x1, y: y1 } = p1;
    const { x: x2, y: y2 } = p2;
    const dx = Math.abs(x1 - x2) ** 2;
    const dy = Math.abs(y1 - y2) ** 2;
    return Math.sqrt(dx + dy);
  }

  function reset() {
    points = [];
    startTime = 0;
    topCtx.beginPath();
  }

  function onMouseMove(e: MouseEvent) {
    const point = {
      x: (e.pageX - offsetLeft) * window.devicePixelRatio,
      y: (e.pageY - offsetTop) * window.devicePixelRatio,
      v: 1,
    };
    points.push(point);
    const currentTime = Date.now();
    let vt = 1;
    if (startTime !== 0) {
      const duration = currentTime - startTime;
      const dx = distance(points[points.length - 2], points[points.length - 1]);
      vt = Math.max(Math.min(dx / duration, lineWidth), 1);
    }
    point.v = vt;
    startTime = currentTime;

    topCtx.save();
    topCtx.lineWidth = lineWidth / vt;
    const { x, y } = points[points.length - 1];
    topCtx.lineTo(x, y);
    topCtx.stroke();
    topCtx.restore();

    if (points.length === 2) {
      const [ { x: x1, y: y1 } ] = points;
      const [ , { x: x2, y: y2 } ] = points;
      if (x1 === x2 && y1 === y2) {
        drawPoint();
        reset();
      }
    }
  }

  function onMouseUp() {
    if (points.length > 0) {
      draw();
    }
    document.removeEventListener('mousemove', onMouseMove);
    document.removeEventListener('mouseup', onMouseUp);
  }

  function onMouseDown() {
    reset();
    document.addEventListener('mousemove', onMouseMove);
    document.addEventListener('mouseup', onMouseUp);
  }

  topCanvas.addEventListener('mousedown', onMouseDown);
}

export default {
  main,
};
